lib/SILGen/SILGenApply.cpp

```diff
@@ -1958,6 +1959,109 @@ buildBuiltinLiteralArgs(SILGenFunction &SGF, SGFContext C,
   return args;
 }
 
+static StringRef transcribe(SILGenFunction &SGF, ASTContext& ctx,
+                            SourceRange exprRange) {
+  // Find the rance of source characters corresponding to the input expression.
+  // If no such range can be found, assume the expression was compiler-generated.
+  auto sourceRange = Lexer::getCharSourceRangeFromSourceRange(ctx.SourceMgr,
+                                                              exprRange);
+  if (sourceRange.isValid()) {
+    return ctx.SourceMgr.extractText(sourceRange);
+  }
+
+  SGF.SGM.diagnose(exprRange.Start, diag::magic_identifier_poor_transcription);
+  return "<compiler-generated>";
+}
+
+static Expr *findArgumentExprWithLabel(DefaultArgumentExpr *defaultArgExpr,
+                                       StringRef expectedLabel) {
+  if (auto argList = defaultArgExpr->getCallerArgs()) {
+    for (auto arg : *argList) {
+      auto label = arg.getInternalLabel().str();
+      if (label.empty()) {
+        label = arg.getLabel().str();
+      }
+      if (label.str() == expectedLabel) {
+        return arg.getExpr();
+      }
+    }
+  }
+
+  return nullptr;
+}
+
+static inline PreparedArguments
+buildTranscriptionLiteral(SILGenFunction &SGF, SGFContext C,
+                          MagicIdentifierLiteralExpr *magicLiteral) {
+  ASTContext &ctx = SGF.getASTContext();
+  auto kindString = MagicIdentifierLiteralExpr::getKindString(
+    magicLiteral->getKind());
+
+  // Get the literal's argument list and verify it has exactly one element.
+  auto argList = magicLiteral->getArgs();
+  if (!argList || argList->size() == 0) {
+    SGF.SGM.diagnose(magicLiteral->getLoc(),
```

Comment:
All of these are semantic checks we should be issuing much earlier. This function should essentially just be extracting the string value and emitting a constant for it.

Reply:
The layering here isn't quite right, yeah. Can you provide some concrete suggestions?